### ISSUE TITLE
composite HealthIndicator that runs them in parallel

### ISSUE THREAD

**Issue by snicoll** [spring-projects/spring-boot#2652]
Right now, invoking the health endpoint lead to a synchronous invocations of all underlying `HealthIndicator` that have been configured. It would be nice to offer an option to run these in parallel instead.

**Comment by philwebb** [comment]
Definitely in the "nice to have" bucket. I imaging that the `health` endpoint isn't hit all that often so performance probably isn't a big deal.

**Comment by snicoll** [comment]
@philwebb I had the idea when I saw #2630

**Comment by fatroom** [comment]
I would also like to have support for this. My project got a bunch of health checks and some of them can run up to few seconds. So in total performance of /health endpoint is terrible with sequential approach.

**Comment by fjcano** [comment]
:+1: 
Some of my projects have 20+ health checks.
For now we use our homebrew solution to do them in parallel.

Would love to see support for parallel health checks in Spring Boot out of the box.
Or at least a way to inject our own `CompositeHealthIndicator`.

**Comment by prismec** [comment]
And it would be nice to include a standard detail information how long a certain health indicator took to execute.

**Comment by erikgreif-acc** [comment]
Real-world example here:

Some services have requirements to respond within a given time limit. This time limit may be set by an external organization, not always under the control of the devs. This kind of timeout gets interpreted as a DOWN state, wakes devs up in the middle of the night, reboots service instances, and all the while provides no context as to which component was experiencing high latency (because there was no HTTP response). 

I have successfully added timeouts per-actuator, but with 20 checks that require timeouts of around 5 seconds (any less would cause alerts for minor network blips), we are asking our caller to wait up to 1:40 for a response. At my org, this is absurd, and nobody wants to permit this behavior.

**Comment by rmie** [comment]
I second everything @erikgreif-acc said.

Sometimes I run into a situation in which a time out occurs but it is unclear which of the components caused it. So I'd like to propose a timeout for the parallel execution after which every unresponsive component is considered/reported as down. To keep the existing behavior the default for this timeout should be infinite.

And while working on it, exporting metrics (number of total, successful and failed executions, sum of execution time for successful executions for each component) would make alerting (and inhibiting) on health related issues easier and would help with dissecting problems faster.

**Comment by jim-mclean** [comment]
+1 for this feature, or for some sort of async checking...

I am surprised that this has not yet been implemented.  We're running Spring Boot Apps on Pivotal (Tanzu?) Cloud Foundry and the out of the box settings in that environment at to consider a HTTP Healthcheck failed after 1s.  For all but the simplest of apps that we have deployed to PCF this is causing unnecessary restarts.

**Comment by wilkinsona** [comment]
https://github.com/spring-projects/spring-boot/issues/25459 is somewhat related to this. I've been wondering about something that runs the health indicators periodically in the background and the health endpoint would then just respond immediately with the current aggregated status.

**Comment by jim-mclean** [comment]
@wilkinsona I think providing framework support for running health indicators in the background and providing the latest aggregate state on HTTP calls would solve this problem pretty holistically.

Initially blocking until there is a response from all indicators seams reasonable.  As does considering an indicator as being DOWN if its invocation has taken longer than some threshold and/or older than some max age.

Also, it may be best to default to the current behavior, but be able to flag HealthIndicator implementations as beaning async with their timeout..  i.e.  @BackgroundCheck(maxTime=2000, maxAge= 3000) or some other existing more suitable annotation.

-

**Comment by antoinemeyer** [comment]
I ended up creating an implementation of async health indicators in this separate repo: https://github.com/antoinemeyer/spring-boot-async-health-indicator
It relies on a `ScheduledThreadPoolExecutor` with an annotation `@AsyncHealth`.

I considered having it only property-driven (`management.health.async.indicator.$INDICATOR-BEAN-NAME.refresh-rate`) or using a different interface (`AsyncHealthIndicator`) but believed an annotation would be more adequate.

I would be very happy to have your feedback on the approach taken. If you think this can be integrated directly into spring-boot-actuator, I'd gladly work on a PR.

**Comment by philwebb** [comment]
Thanks very much for the pull-request offer @antoinemeyer. It's really great to see that you've been able to add support in an additional library. We're currently not sure if we want to offer async support in the way that you've implemented it, or as suggested in #25459. Unfortunately, we're currently investing most of our time in Spring Boot 3.0 work so it's unlikely that we'll be able to make the decision anytime soon. I think it's best we wait until Spring Boot 3.0 is out before looking at this. That probably means that a pull-request would be a bit premature, and it would be best to consider things again when we're planning Spring Boot 3.1.

**Comment by antoinemeyer** [comment]
Thanks @philwebb for the feedback (once again).
I am looking forward to see what solution you will end up adopting for this problem. In the meantime I'll rely on (and improve) that library and hope it can benefit others!

**Comment by unnisathya92** [comment]
hi @antoinemeyer, Do you have a working example for the [aysnc-health-proj](https://github.com/antoinemeyer/spring-boot-async-health-indicator)? It would really be helpful

**Comment by antoinemeyer** [comment]
> hi @antoinemeyer, Do you have a working example for the [aysnc-health-proj](https://github.com/antoinemeyer/spring-boot-async-health-indicator)? It would really be helpful

hi @unnisathya92 ! Sure, here is a starter with asynchronous health indicators: [spring-async-health-indicator-demo.zip](https://github.com/spring-projects/spring-boot/files/7781791/spring-async-health-indicator-demo.zip)

**Comment by kcshettar** [comment]
+1 for the feature. My application has three health check components. 

Component 1: <100ms response time
Component 2: <400ms response time
Component 3: <400ms response time

Would love to see the `/actuator/health` endpoint to respond within 400ms (worst case).

I am rendering the health details on react application (Home Page) which takes 100ms+400ms+400ms = 900ms right now. 900ms page load time is acceptable right now. It would be nice if the Home Page can load within 400ms.

We could implement the parallel processing logic as a temporary solution. But, I am sure this feature will come out of the box some day under `spring-boot-starter-actuator`.

**Comment by ahoehma** [comment]
Any news here? In my app I have multiple db's, so multiple db-pools and so multiple health checks.

Now I see something like this:

Health contributor org.springframework.boot.actuate.jdbc.DataSourceHealthIndicator (db/creationCacheDbDataSource) took 269391ms to respond

This is really bad :)

Is there something I can do to pimp my (db) health checks to be little bit more smart regarding resources/runtime etc?

**Comment by snicoll** [comment]
@ahoehma not sure you're looking in the right direction. Running them in parallel should improve things a bit, but certainly not to the extend of what you're experiencing now. In other words, please review your indicator and backend configuration.

**Comment by snicoll** [comment]
That, and we don't use the issue tracker for discussions. You may want to chat with the community on Gitter. You may also be interested in [endpoint caching](https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.caching).

**Comment by antoinemeyer** [comment]
> Hopefully this is helpful to someone else here or in #25459

FYI I updated [spring-boot-async-health-indicator](https://github.com/antoinemeyer/spring-boot-async-health-indicator) under version `boot3-1.3` for spring boot 3 support.
